pulsar/schema.go

```diff
@@ -66,13 +67,21 @@ type SchemaInfo struct {
 	Schema     string
 	Type       SchemaType
 	Properties map[string]string
+	hashVal    atomic.Uint64
 }
 
-func (s SchemaInfo) hash() uint64 {
+func (s *SchemaInfo) hash() uint64 {
+	oldHash := s.hashVal.Load()
+	if oldHash != 0 {
+		return oldHash
+	}
+
 	h := maphash.Hash{}
 	h.SetSeed(seed)
 	h.Write([]byte(s.Schema))
-	return h.Sum64()
+	newHash := h.Sum64()
+	s.hashVal.CompareAndSwap(oldHash, newHash)
```

Comment:
We can use `sync.Once` to ensure `hashVal` just set value on first call time.

``` go
	s.onc.Do(func() {
		h := maphash.Hash{}
		h.SetSeed(seed)
		h.Write([]byte(s.Schema))
		s.hashVal = h.Sum64()
	})
	return s.hashVal
```

Reply:
I think `atomic` is enough, 'cause in sync.Once, there is an `atomic` and a `mutex`, when we read the hash, it will `atomic.LoadUint32(&o.done)` first, the costs are the same.

``` go
type Once struct {
	// done indicates whether the action has been performed.
	// It is first in the struct because it is used in the hot path.
	// The hot path is inlined at every call site.
	// Placing done first allows more compact instructions on some architectures (amd64/386),
	// and fewer instructions (to calculate offset) on other architectures.
	done uint32
	m    Mutex
}

func (o *Once) Do(f func()) {
	// Note: Here is an incorrect implementation of Do:
	//
	//	if atomic.CompareAndSwapUint32(&o.done, 0, 1) {
	//		f()
	//	}
	//
	// Do guarantees that when it returns, f has finished.
	// This implementation would not implement that guarantee:
	// given two simultaneous calls, the winner of the cas would
	// call f, and the second would return immediately, without
	// waiting for the first's call to f to complete.
	// This is why the slow path falls back to a mutex, and why
	// the atomic.StoreUint32 must be delayed until after f returns.

	if atomic.LoadUint32(&o.done) == 0 {
		// Outlined slow-path to allow inlining of the fast-path.
		o.doSlow(f)
	}
}
```

Reply:
But, Using `sync.once` is more intuitive.

And, In multi-thread, your implementation of this is not guaranteed to be hashed only once.

Reply:
@shibd I have update to sync.Once since there is no atomic.Uint64 in go 1.18